src/solver/loss/backward_hinge_embedding_loss.cpp

```diff
@@ -0,0 +1,198 @@
+                                std::string kernel_name,
+                                KernelBuildParameters build_params) {
+    while(localsize.size() < 3)
+        localsize.push_back(1);
+    while(gridsize.size() < 3)
+        gridsize.push_back(1);
+    for(int i = 0; i < localsize.size(); ++i)
+        gridsize[i] = AlignUp(gridsize[i], localsize[i]);
+    return KernelInfo{
+        build_params.GenerateFor(kbp::HIP{}), localsize, gridsize, kernel_file, kernel_name};
+};
+
+bool HingeEmbeddingLossBwd::IsApplicable(
+    const ExecutionContext& /*context*/,
+    const miopen::loss::HingeEmbeddingLossBwdProblemDescription& problem) const
+{
+    if(problem.GetIDesc().GetSize() > 5)
+        return false;
+    return true;
+}
+
+ConvSolution HingeEmbeddingLossBwd::GetSolution(
+    const ExecutionContext& context,
+    const miopen::loss::HingeEmbeddingLossBwdProblemDescription& problem) const
+{
+    std::ignore = context;
+
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    auto in_dtype     = miopen::GetDataType(problem.GetIDesc().GetType());
+    auto dtype        = problem.GetdIDesc().GetType();
+    auto target_dtype = miopen::GetDataType(problem.GetTDesc().GetType());
+
+    const auto build_params = KernelBuildParameters{
+        {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+        {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+        {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+        {"IN_OUT_TYPE", in_dtype == "bfloat16" ? "ushort" : in_dtype},
+        {"TARGET_TYPE", target_dtype},
+        {"LOCAL_SIZE", LOCAL_SIZE},
+    };
+
+    result.construction_params.push_back(make_hip_kernel({LOCAL_SIZE},
+                                                         {problem.GetIDesc().GetElementSize()},
+                                                         "MIOpenHingeEmbeddingLoss.cpp",
+                                                         "HingeEmbeddingLossBwd",
+                                                         build_params));
+
+    result.invoker_factory = [](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+            decltype(auto) kernel = handle_.Run(kernels.front());
+            decltype(auto) params = raw_params.CastTo<miopen::loss::BwdInvokeParams>();
+            auto I_tv             = get_inner_expanded_tv(deref(params.iDesc));
+            auto T_tv             = get_inner_expanded_tv(deref(params.tDesc));
+            auto dO_tv            = get_inner_expanded_tv(deref(params.dODesc));
+
+            kernel(params.i,
+                   params.t,
+                   params.dO,
+                   params.dI,
+                   params.margin,
+                   params.divisor,
+                   I_tv,
+                   T_tv,
+                   dO_tv);
+        };
+    };
+
+    return result;
+}
+
+bool HingeEmbeddingLossUnreducedBwd::IsApplicable(
```

Comment:
HingeEmbeddingLossBwd
Please separate the HingeEmbeddingLossBwdand HingeEmbeddingLossUnreducedBwdfiles.

Reply:
I have splited each kernel to a separate file